### PR TITLE
[CSHARP-969] - OpenTelemetry tracing RFC

### PR DESCRIPTION
**Pull request by simaoribeiro**
As discussed in https://datastax-oss.atlassian.net/browse/CSHARP-969?focusedCommentId=55554, this document describes a proposal for the inclusion of tracing for C# driver that follows the OpenTelemetry specification.

It was based on the template available in the Rust lang repository and should be used as a support guide for the reading of the proposal
https://github.com/rust-lang/rfcs/blob/master/0000-template.md

In the repository below, there is an incomplete POC developed by our team that can be viewed when analyzing the technical parts of the proposal:
https://github.com/simaoribeiro/csharp-driver/pull/2

### CODE REVIEW

File: doc/features/open-telemetry/tracing-rfc.md [hunk 1]
```diff
@@ -0,0 +1,414 @@
+- Feature Name: Add OpenTelemetry Traces
```
Comment (simaoribeiro) [inline doc/features/open-telemetry/tracing-rfc.md]: TODO: Include team in the commit when squashing and rebasing

File: doc/features/open-telemetry/tracing-rfc.md [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+
+Cassandra has a community developed package available in the [opentelemetry-dotnet-contrib](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.Cassandra) that exports metrics but not traces. This proposal to include traces in the native Cassandra C# driver will allow the developers to have access to database operations when analyzing the requests that are handled in their systems when it includes Cassandra calls.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## [Traces](https://opentelemetry.io/docs/concepts/signals/traces/)
+
+As mentioned in [*Motivation*](#motivation), traces allows the developers to understand the full "path" a request takes in the application and navigate through a microservice architecture. Traces include [Spans](https://opentelemetry.io/docs/concepts/signals/traces/#spans) which are unit of works or operation in the ecosystem that include the following information:
```
Comment (joelfoliveira) [inline doc/features/open-telemetry/tracing-rfc.md]: 💅 related with this statement "and navigate through a microservice architecture". Tracing is still valuable in non microservices architecture. So maybe don't bind tracing to that architecture only.

File: doc/features/open-telemetry/tracing-rfc.md [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+| db.name | The keyspace name in Cassandra. | string | true | *keyspace in use* |
+| db.statement | The database statement being executed. | string | false | *database statement in use* |
+| server.address | Name of the database host. | string | true | e.g.: example.com; 10.1.2.80; /tmp/my.sock |
+| server.port | Server port number. Used in case the port being used is not the default | int | false | e.g.: 9445 |
+
+## Usage
+
+### Package installation
+
+The OpenTelemetry implementation is included in the package `CassandraCSharpDriver.OpenTelemetry`.
```
Comment (joelfoliveira) [inline doc/features/open-telemetry/tracing-rfc.md]: Since we are referring to a proposal for future implementation, we should probably rephrase it differently:

```suggestion
The OpenTelemetry implementation will be included in the package `CassandraCSharpDriver.OpenTelemetry`.
```
Does it make sense?

File: doc/features/open-telemetry/tracing-rfc.md [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+
+### Exporting Cassandra activity
+
+The extension method `AddOpenTelemetryInstrumentation()` is available in the cluster builder, so the activity can be exported for database operations:
```
Comment (joelfoliveira) [inline doc/features/open-telemetry/tracing-rfc.md]: Also related to this [comment](https://github.com/simaoribeiro/csharp-driver/pull/1/files#r1441584971) but using the present tense creates some confusion if this is available right now or if it is referring to the proposed change.

File: doc/features/open-telemetry/tracing-rfc.md [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+
+```csharp
+var cluster = Cluster.Builder()
+    .AddContactPoint("127.0.0.1")
+    .WithSessionName("session-name")
+    .AddOpenTelemetryInstrumentation()
```
Comment (joelfoliveira) [inline doc/features/open-telemetry/tracing-rfc.md]: The open-telemetry-contrib for [cassandra](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.Cassandra) already defines a `.WithOpenTelemetryMetrics()`.  By naming this `.AddOpenTelemetryInstrumentation()` couldn't this mistakenly imply that it will enable all the instrumentation for Cassandra (metrics, tracing)?

File: doc/features/open-telemetry/tracing-rfc.md [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+}
+
+public void OnSpeculativeExecution(Host host, long delay)
+{
+}
+```
+
+### Public API
+
+The `Builder` will include a new method named `WithTracer` that will include an `IDriverTrace` instance being passed as parameter. This method can be used by anyone to pass their tracer implementation and will be used by the extension method `AddOpenTelemetryInstrumentation()` mentioned in the previous section.
```
Comment (joelfoliveira) [inline doc/features/open-telemetry/tracing-rfc.md]: What is the use case for someone to provide their own `IDriverTrace` implementation?

**Review by joelfoliveira — APPROVED** [review]

File: doc/features/open-telemetry/tracing-rfc.md (hunk 6 above)
Comment (simaoribeiro) [reply doc/features/open-telemetry/tracing-rfc.md]: I would say, if someone wants to implement their own instrumentation like not using OTEL or prefer to create their own activity with their own attributes, it would make sense.
Also, using IDriverTracer is a way to separate the interface from its implementation without the use of a provider builder, just like Cassandra.AppMetrics implementation has

File: doc/features/open-telemetry/tracing-rfc.md (hunk 5 above)
Comment (simaoribeiro) [reply doc/features/open-telemetry/tracing-rfc.md]: Good question. Metrics are not being exported or created by any standard and, in [future possibilities](https://github.com/simaoribeiro/csharp-driver/blob/793dfec7b6262208094b14c133b42871060e3780/doc/features/open-telemetry/tracing-rfc.md#future-possibilities) I mentioned that metrics could be exported from OTEL in the future.
However, it makes sense that the extension method should have a more explicit name. Maybe `AddOpenTelemetryTraceInstrumentation()`? What do you say?

File: doc/features/open-telemetry/tracing-rfc.md (hunk 2 above)
Comment (simaoribeiro) [reply doc/features/open-telemetry/tracing-rfc.md]: Done, please review

File: doc/features/open-telemetry/tracing-rfc.md (hunk 3 above)
Comment (simaoribeiro) [reply doc/features/open-telemetry/tracing-rfc.md]: Done, please review

File: doc/features/open-telemetry/tracing-rfc.md (hunk 4 above)
Comment (simaoribeiro) [reply doc/features/open-telemetry/tracing-rfc.md]: Done, please review

File: doc/features/open-telemetry/tracing-rfc.md (hunk 5 above)
Comment (simaoribeiro) [reply doc/features/open-telemetry/tracing-rfc.md]: Done, please review

File: doc/features/open-telemetry/tracing-rfc.md [hunk 7]
```diff
@@ -0,0 +1,415 @@
+- Feature Name: Add OpenTelemetry Traces
+- Start Date: 2023-12-27
+
```
Comment (ruiqbarbosa) [inline doc/features/open-telemetry/tracing-rfc.md]: One topic that seems not be covered in this RFC is the [IStatement.IsTracing](https://github.com/simaoribeiro/csharp-driver/blob/bc6863efd17a01c26c46369248205c1a35a5a6e0/src/Cassandra/IStatement.cs#L53) property. 

I am highlighting this because I remember that this property could lead to certain doubts with the OpenTelemetry implementation.

Is it not a concern anymore?

File: doc/features/open-telemetry/tracing-rfc.md (hunk 2 above)
Comment (joelfoliveira) [reply doc/features/open-telemetry/tracing-rfc.md]: looks good.

File: doc/features/open-telemetry/tracing-rfc.md (hunk 3 above)
Comment (joelfoliveira) [reply doc/features/open-telemetry/tracing-rfc.md]: 👍

File: doc/features/open-telemetry/tracing-rfc.md (hunk 4 above)
Comment (joelfoliveira) [reply doc/features/open-telemetry/tracing-rfc.md]: 👍

File: doc/features/open-telemetry/tracing-rfc.md (hunk 5 above)
Comment (joelfoliveira) [reply doc/features/open-telemetry/tracing-rfc.md]: I agree with `AddOpenTelemetryTraceInstrumentation` in order to prevent miss judgement around what that is doing. Unless the implementation of other instrumentation features are planned for the "near" future.

Nevertheless, their input on this is going to be more valuable to understand the best naming.

File: doc/features/open-telemetry/tracing-rfc.md (hunk 6 above)
Comment (joelfoliveira) [reply doc/features/open-telemetry/tracing-rfc.md]: Ok 👍

**Review by JoaoRodriguesGithub — COMMENTED** [review]
Lgtm

**Review by ruiqbarbosa — APPROVED** [review]

File: doc/features/open-telemetry/tracing-rfc.md [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+`db.statement` is optional given that this attribute may contain sensitive information:
+
+| Attribute  | Description  | Type | Required | Supported Values|
+|---|---|---|---|---|
+| span.kind | Describes the relationship between the Span, its parents, and its children in a Trace. | string | true | client |
+| db.system | An identifier for the database management system (DBMS) product being used. | string | true | cassandra |
+| db.name | The keyspace name in Cassandra. | string | true | *keyspace in use* |
+| db.statement | The database statement being executed. | string | false | *database statement in use* |
+| server.address | Name of the database host. | string | true | e.g.: example.com; 10.1.2.80; /tmp/my.sock |
+| server.port | Server port number. Used in case the port being used is not the default | int | false | e.g.: 9445 |
```
Comment (lpcouto) [inline doc/features/open-telemetry/tracing-rfc.md]:
```suggestion
| server.port | Server port number. Used in case the port being used is not the default. | int | false | e.g.: 9445 |
```

File: doc/features/open-telemetry/tracing-rfc.md [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+
+## Using OpenTelemetry.SemanticConventions package
+
+The [semantic conventions](https://opentelemetry.io/docs/specs/semconv/) are a fast evolving reference that "define a common set of (semantic) attributes which provide meaning to data when collecting, producing and consuming it.".\
+As it changes can be hard to follow, the .NET project includes a package named [`OpenTelemetry.SemanticConventions`](https://www.nuget.org/packages/OpenTelemetry.SemanticConventions/1.0.0-rc9.9) that maps the attributes defined in the conventions to a .NET project. Using this package will allow the Cassandra project to have its tracing attributes up-to-date to the conventions with less maintenance, however, as it's still marked as non-stable (current version is `1.0.0-rc9.9`), its inclusion is not included in this proposal.
```
Comment (lpcouto) [inline doc/features/open-telemetry/tracing-rfc.md]:
```suggestion
As its changes can be hard to follow, the .NET project includes a package named [`OpenTelemetry.SemanticConventions`](https://www.nuget.org/packages/OpenTelemetry.SemanticConventions/1.0.0-rc9.9) that maps the attributes defined in the conventions to a .NET project. Using this package will allow the Cassandra project to have its tracing attributes up-to-date to the conventions with less maintenance, however, as it's still marked as non-stable (current version is `1.0.0-rc9.9`), it is not included in this proposal.
```

File: doc/features/open-telemetry/tracing-rfc.md [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+[unresolved-questions]: #unresolved-questions
+
+- TBD
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Traces
+
+### Include missing Recommended attributes
```
Comment (lpcouto) [inline doc/features/open-telemetry/tracing-rfc.md]:
```suggestion
### Include missing recommended attributes
```

**Review by lpcouto — APPROVED** [review]

File: doc/features/open-telemetry/tracing-rfc.md [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+{
+    (...)
+    
+    ObserverFactoryBuilder = observerFactoryBuilder ?? (MetricsEnabled ? (IObserverFactoryBuilder)new MetricsObserverFactoryBuilder() : new NullObserverFactoryBuilder());
+
+    (...)
+}
+```
+
+As it is necessary to have multiple observers being triggered on driver'actions, this proposal includes a composite pattern that will aggregate multiple observer instances for that work.
```
Comment (brmagadutra) [inline doc/features/open-telemetry/tracing-rfc.md]:
```suggestion
As it is necessary to have multiple observers being triggered on driver's actions, this proposal includes a composite pattern that will aggregate multiple observer instances for that work.
```

File: doc/features/open-telemetry/tracing-rfc.md [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+As so, the instantiation of the `IRequestResultHandler` in `RequestHandler` class will also change to include the statement as parameter_
+
+```csharp
+_requestResultHandler = new TcsMetricsRequestResultHandler(_requestObserver, statement);
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Although the implementation suggested in this document tries to minimize changing the core package, the inclusion of new observers being called will increase the complexity and the performance of the driver.
```
Comment (brmagadutra) [inline doc/features/open-telemetry/tracing-rfc.md]: Will it increase the performance or it will decrease it?

**Review by brmagadutra — APPROVED** [review]

**Review by JoaoRodriguesGithub — APPROVED** [review]

File: doc/features/open-telemetry/tracing-rfc.md (hunk 1 above)
Comment (simaoribeiro) [reply doc/features/open-telemetry/tracing-rfc.md]: Co Authored and reviwed-by included in git message
